Fixes of Zim file does not download / pending forever

MohitMaliFtechiz opened the pull request:
Fixes #3441 

We encountered an issue where the download remained pending indefinitely if we canceled it from the notification. This happened because when we canceled the download by clicking the notification cancel button, it kept a reference in the fetch system. Consequently, if we tried to download the same file again, fetch created a new download ID, resulting in a conflict with the existing instance of the file. As a result, the download did not start.

However, when we canceled the download within the app itself, it deletes the instance from fetch, which automatically deletes the temporary file from storage.

To resolve this issue, we have made a change. Instead of using the CANCEL action, we now use the DELETE action. So, if the user cancels the download from the notification, it will function the same way as if they have canceled it from within the application.


https://github.com/kiwix/kiwix-android/assets/34593983/b8d130f5-d58c-44de-a875-707d31d70253

gouri-panda reviewed (dismissed):
@MohitMaliDeveloper @kelson42 LGTM! @MohitMaliDeveloper Thanks! Would be great if we added tests.

kelson42 commented:
@MohitMaliDeveloper strongly disagree with this PR because what you have developed is a pause/resume feature without calling it appropriatly and hidding it behind the cancel feature. Basically a failed attempt to fix #2093.

Naming things properly is the highest challeng in software engineering, by not doing so you have created new bugs…. And actually the PR does not match the bug reported first. @gouri-panda This PR ahould not have been approved.

An example of bug: user download 1GB of data for a zim then want to cancel. He cancels… but now one temporary data of 1GB is on the disk and he has no way to recover that space. If he do that a few times, he can fill the disk with twmporary data!

Now, @MohitMaliDeveloper please:
* Implement a fix for #3441 and only that
* Move the pause resume code part in a PR (or keep that one) with purpose to implement #2093, but first propose a UI change so the UX is proporr.

@MohitMaliDeveloper Does all of this make sense or do I’m missing aomething?

MohitMaliFtechiz commented:
> strongly disagree with this PR because what you have developed is a pause/resume feature without calling it appropriatly and hidding it behind the cancel feature.
Naming things properly is the highest challeng in software engineering, by not doing so you have created new bugs…. And actually the PR does not match the bug reported first.

@kelson42, Yes, we have implemented the pause/resume feature. I properly updated the PR description, taking reference from Gouri's comment to reduce extra network costs. However, I apologize for not updating the PR title. The main issue we addressed was when canceling a download from the notification, we used the fetch cancel function, which cancels the download. If we then attempt to download the same item again, the file is already present in storage and we end up with two download IDs for the same file in fetch. On the other hand, when canceling the download from the application, it deletes the download ID from fetch, which automatically removes the corresponding file from storage.


> user download 1GB of data for a zim then want to cancel. He cancels… but now one temporary data of 1GB is on the disk and he has no way to recover that space. If he do that a few times, he can fill the disk with temporary data!

I agree with you. The memory occupied by the canceled download is not recoverable from within the application itself. The user can only recover that memory by manually deleting the file using a file manager.

> Move the pause resume code part in a PR (or keep that one) with purpose to implement https://github.com/kiwix/kiwix-android/issues/2093, but first propose a UI change so the UX is proporr.

In the current UI, when the user clicks on the cancel button, we show a dialog to confirm the cancellation of the download. We can enhance this dialog by adding a new button called "Pause." When the user clicks on the Pause button, the download will be paused, and we will update the UI to display a pause button instead of the stop button. If the user clicks on the pause button, we can show another dialog with two options:
a) Resume the download.
b) Cancel the download, which will delete the files from storage.
However, this approach has a drawback. For instance, if the user is at the bottom or in the middle of the download list and pauses a download, then closes the application. When they want to resume the download later, they would need to find that particular zim file in the list. This approach can be frustrating for the user. To address this issue, we have two additional approaches:

a) Show Paused Downloads at the Top:
We can display the paused downloads at the top of the list, making it easier for the user to locate and resume them.

b) Create a Separate Screen for Paused Downloads:
We can introduce a separate screen dedicated to displaying paused downloads. This way, the user can easily access and manage their paused downloads without having to search for them in the main download list.

kelson42 commented:
> The main issue we addressed was when canceling a download from the notification, we used the fetch cancel function, which cancels the download. If we then attempt to download the same item again, the file is already present in storage and we end up with two download IDs for the same file in fetch. On the other hand, when canceling the download from the application, it deletes the download ID from fetch, which automatically removes the corresponding file from storage.

It's mandatory to always delete temporary file if download is cancel. If not always done, this is a bug and this is what has to be fixed here.

MohitMaliFtechiz commented:
> It's mandatory to always delete temporary file if download is canceled. If not always done, this is a bug and this is what has to be fixed here.

@kelson42, I have done the changes as you requested also updated the PR description.

kelson42 commented:
@gouri-panda can you èlease review again?

gouri-panda commented:
>An example of bug: user download 1GB of data for a zim then want to cancel. He cancels… but now one temporary data of 1GB is on the disk and he has no way to recover that space. If he do that a few times, he can fill the disk with twmporary data!

@kelson42 Sorry, I didn't see this edge case. You made a good valid point. Storage is also our priority too. I agree with @MohitMaliDeveloper too, we should have pause, resume and cancel button use cases too. pause will allow to stop downloading files for a while. resume will allow start downloading from the previous state and cancel will allow stop the downloading  & delete the whole file from the storage.

>We can display the paused downloads at the top of the list, making it easier for the user to locate and resume them.

@MohitMaliDeveloper I think we have this ticket before. Perhaps this is a good time to remove the dependency of the fetch library. Also, we'll benefit the duplicate id problem you previously discussed.

kelson42 commented:
Not in favour of replacing Fetch now.... not a good timing for such a big change.

MohitMaliFtechiz commented:
@kelson42, I have created a PR https://github.com/kiwix/kiwix-android/pull/3459 with these pause/resume download changes.
